keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,788 @@
+
+- Since the current discovery separated by group-version is already
+  GA, removal of the endpoint will be difficult. This KEP will solely
+  focus on introducing the new aggregated endpoint and will not cover
+  deprecation.
+
+## Proposal
+
+We are proposing a new endpoint `/discovery/v1` as an aggregated
+endpoint for all discovery documents. Discovery documents can
+currently be found under `apis/<group>/<version>` and `/api/v1` for
+the legacy group version. This discovery endpoint will support
+publishing an ETag so clients who already have the latest version of
+the aggregated discovery can avoid redownloading the document.
+
+We will add a new controller responsible for aggregating the discovery
+documents when a resource on the cluster changes. There will be no
+conflicts when aggregating since each discovery document is
+self-contained.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!-- What are the caveats to the proposal? What are some important
+details that didn't come across above? Go in to as much detail as
+necessary here. This might be a good place to talk about core concepts
+and how they relate. -->
+
+### Risks and Mitigations
+
+<!-- What are the risks of this proposal, and how do we mitigate?
+Think broadly. For example, consider both security and how this will
+impact the larger Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!-- This section should contain enough information that the specifics
+of your change are understandable. This may include API specs (though
+not always required) or even code snippets. If there's any ambiguity
+about HOW your proposal will be implemented, this is the place to
+discuss them. -->
+
+We will expose a endpoint `/discovery` that will support JSON, and
+protobuf and gzip compression. The endpoint will serve the aggregated
+discovery document for all types that a Kubernetes cluster supports.
+
+### API
+
+The contents of this endpoint will be an `APIGroupList`, which is the
+same type that is returned in the discovery document at `/apis`. The
+`APIGroupList` contains a list of `APIGroup` types which includes a
+list of
+[`GroupVersionForDiscovery`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1071)
+types. We will modify this `GroupVersionForDiscovery` type to include
+a list of
+[`APIResource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1080).
+This list of `APIResource` is what is currently published at the
+`/apis/<group>/version` endpoint and is what will be aggregated into
+the new endpoint that we publish.
+
+The endpoint will also publish an ETag calculated based on a hash of
+the data for clients.
+
+### Aggregation
+
+For the aggregation layer on the server, a new controller will be
```

Comment:
This will likely need to have an input to readiness and you'll need to filter the discovery results by applicable group/version.

Reply:
Yeah it's real bad if the first few clients get an empty doc because this controller hasn't run yet, but it's also bad to delay the cluster startup.

Reply:
If api server signals it is ready even when discovery aggregation is not completed, this will not only affects clients but also causes flakiness of some e2e tests depends on discovery?

Reply:
> If api server signals it is ready even when discovery aggregation is not completed, this will not only affects clients but also causes flakiness of some e2e tests depends on discovery?

More importantly, that causes bad behavior of the namespace lifecycle controller.

Reply:
ah yeah that's another thing to consider, how does this affect that controller.
* will clients be able to know that e.g. a specific GV's backing apiserver hasn't been responsive for a while?
* if an apiserver restarts in that situation, how does it build the cache?

Reply:
Added that this will be an input to readiness.

For the controllers, afaik the behavior is based on the rest mapper. I've talked to @apelisse and I think we could keep the rest mapper behavior for controllers untouched and still have them use the old unaggregated endpoints. A single controller generally only cares about a very small subset of resources and should not need the entire discovery document. Since we will not be removing the old endpoint, we can optimize the existing controllers to only query for gvs that they need and have a new restmapper that uses the aggregated document.